### PR TITLE
Review sample preparation process, add augmentations, bug fixes.

### LINKED ISSUE #68: NaN values in input images
## Current behavior
Not a number (NaN) and infinite (inf) values in images are also stored in HDFS files. Those values will affect the loss calculation, resulting in a NaN loss, therefore not learning anything.  
Here's an example of the output (warning from numpy) and training and validation loss are set to nan because of if:  
```shell
Start:

Epoch 0/49
--------------------
/home/ec2-user/miniconda3/lib/python3.6/site-packages/numpy/core/_methods.py:32: RuntimeWarning: invalid value encountered in reduce
  return umr_minimum(a, axis, None, out, keepdims, initial)
Training Loss: nan
Training iou: 0.9478
Training precision: 0.8983
Training recall: 0.9478
Training f1-score: 0.9224
Validation Loss: nan
Validation iou: 0.9395
Validation precision: 0.8827
Validation recall: 0.9395
Validation f1-score: 0.9102
Current elapsed time 27m 17s  
...
```  
## Expected behavior  
NaN values in images should be written as 0 in the hdfs files.  
Infinite values should be reduced.

### PR DESCRIPTION
**Pull request by remtav**
Sample prep changes:
- clip raster to gpkg when gpkg's extent is smaller than raster. Requires geopandas (can we find a work around without geopandas?)
- always pad sat images with nan values when possible, then convert those values to 0 during training
- always pad map images (labels) with ignore_index value to keep any new "background" values (eg if padding with 0 value) from being created, as was done previously

Augmentations:
- Add RadiometricTrim with random trim value inside an inputted range
- Add RandomCrop with an inputted target_size. Goal: decrease amount of duplicaton at sample prep step and maximise effect of seing same sample with small offset (previously done with large overlap at sample prep step)

Fixes #68, #131, #127, #115

### CODE REVIEW

File: conf/config.yaml [hunk 1]
```diff
@@ -43,7 +43,7 @@ sample:
   prep_csv_file: /path/to/csv/file.csv
   val_percent: 5 # Percentage of validation samples created from train set (0 - 100)
   overlap: 25
-  sampling: {'method':['min_annotated_percent'], 'map': 0, '0':0, '1':0, '2':0} # One of class_proportion, annotated_percent
+  sampling_method: {'min_annotated_percent': 0, 'class_proportion': {'0':0, '1':0}} # FIXME. # One of class_proportion or min_annotated_percent
```
Comment (felegare) [inline conf/config.yaml:46]: Tu n'as pas de problèmes même si tu ne mets pas toutes les clés et les valeurs de toutes les classes dans le dictionnaire 'class_proportion'?

File: inference.py [hunk 2]
```diff
@@ -115,10 +134,11 @@ def sem_seg_inference(model, nd_array, overlay, chunk_size, num_classes, device,
...
                                                           overlay=overlay))
-            if debug:
-                output_counts_PIL = Image.fromarray(output_counts.astype(np.uint8), mode='L')
-                output_counts_PIL.save(output_path.joinpath(f'output_counts.png'))
-                tqdm.write(f'Dividing array according to output counts...\n')
+
+            # if debug:
+            #     output_counts_PIL = Image.fromarray(output_counts.astype(np.uint8), mode='L')
+            #     output_counts_PIL.save(output_path.joinpath(f'output_counts.png'))
+            #     tqdm.write(f'Dividing array according to output counts...\n')
```
Comment (mpelchat04) [inline inference.py:141]: Is this code commented because debug mode at inference is deprecated?

File: inference.py [hunk 3]
```diff
@@ -273,7 +293,7 @@ def main(params):
         else:
             img_dir = Path(img_dir_or_csv)
             assert img_dir.is_dir(), f'Could not find directory "{img_dir_or_csv}"'
-            list_img_paths = sorted(img_dir.glob('*.tif'))  # FIXME: what if .tif is in caps (.TIF) ?
+            list_img_paths = sorted(img_dir.glob('*.tif'))  # FIXME: what if .tif is in caps (.TIF) ? linux is case-sensitive
```
Comment (mpelchat04) [inline inference.py:296]: `lower(str)` maybe? This would prevent case-sensitivity related problems.

File: utils/augmentation.py [hunk 4]
```diff
@@ -1,51 +1,170 @@
...
+        if type == 'radiometric':
+            radiom_trim_range = get_key_def('radiom_trim_range', params['training']['augmentation'], None)
+            brightness_contrast_range = get_key_def('brightness_contrast_range', params['training']['augmentation'], None)
+            noise = get_key_def('noise', params['training']['augmentation'], None)
+
+            if radiom_trim_range:  # Contrast stretching
+                lst_trans.append(RadiometricTrim(range=radiom_trim_range))  # FIXME: test this. Assure compatibility with CRIM devs (don't trim metadata)
+
+            if brightness_contrast_range:
+                # lst_trans.append()
```
Comment (mpelchat04) [inline utils/augmentation.py:39]: You could `raise NotImplementedError`.

File: utils/geoutils.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        minx, miny, maxx, maxy = src.bounds  # ouest, nord, est, sud
+
+    # Create a bounding box with Shapely # TODO: can we use only fiona and rasterio for this? No geopandas and shapely ideally.
+    bbox = box(minx, miny, maxx, maxy)
+
+    # Insert the bbox into a GeoDataFrame
+    geo = gpd.GeoDataFrame({'geometry': bbox}, index=[0]) #, crs=gpkg_crs['init'])
+
+    # Re-project into the same coordinate system as the raster data
+    # geo = geo.to_crs(crs=raster.crs.data)
```
Comment (mpelchat04) [inline utils/geoutils.py:101]: I'd like that very much.

File: inference.py [hunk 6]
```diff
@@ -331,11 +349,10 @@ def main(params):
                         "global configuration requested metadata mapping onto loaded samples, but raster did not have available metadata"
                     metadata = read_parameters(img['meta'])
 
-                if debug:
-                    _tqdm.set_postfix(OrderedDict(img_name=img_name,
-                                                  img=np_input_image.shape,
-                                                  img_min_val=np.min(np_input_image),
-                                                  img_max_val=np.max(np_input_image)))
+                _tqdm.set_postfix(OrderedDict(img_name=img_name,
```
Comment (CharlesAuthier) [inline inference.py:352]: why this is not in debug anymore?

File: utils/metrics.py [hunk 7]
```diff
@@ -71,7 +71,7 @@ def report_classification(pred, label, batch_size, metrics_dict, ignore_index=-1
     return metrics_dict
 
 
-def iou(pred, target, batch_size, metrics_dict):
+def iou(pred, target, batch_size, metrics_dict):  # TODO: activate IoU
```
Comment (CharlesAuthier) [inline utils/metrics.py:74]: what do you mean by activate?

File: utils/CreateDataset.py [hunk 8]
```diff
@@ -105,9 +105,10 @@ def __getitem__(self, index):
             if meta_idx != -1:
                 metadata = self.metadata[meta_idx]
                 metadata = eval(metadata) if isinstance(metadata, str) else metadata  # FIXME!!! must provide dtype
-            sample_indices = hdf5_file["sample_indices"][index, ...]  # Only useful fur debugging
+            sat_img_dtype = hdf5_file["sat_img_dtype"][index, ...][0]  # FIXME: replace
+            # sample_indices = hdf5_file["sample_indices"][index, ...]  # Only useful fur debugging
```
Comment (CharlesAuthier) [inline utils/CreateDataset.py:110]: Maybe create a `if debugging`, just an idea but that can be useful later I think.

File: utils/metrics.py [hunk 9]
```diff
@@ -64,14 +64,16 @@ def report_classification(pred, label, batch_size, metrics_dict, ignore_index=-1
             metrics_dict['recall_' + key].update(class_score[key]['recall'], batch_size)
             metrics_dict['fscore_' + key].update(class_score[key]['f1-score'], batch_size)
 
+
     metrics_dict['precision'].update(class_report['weighted avg']['precision'], batch_size)
     metrics_dict['recall'].update(class_report['weighted avg']['recall'], batch_size)
     metrics_dict['fscore'].update(class_report['weighted avg']['f1-score'], batch_size)
+    metrics_dict = iou(pred, label, batch_size, metrics_dict)
```
Comment (CharlesAuthier) [inline utils/metrics.py:71]: Just maybe we need to talk about the implementation and how it's call, make it similar to the lines before.

File: utils/visualization.py [hunk 10]
```diff
@@ -111,23 +111,24 @@ def vis(params, input, output, vis_path, sample_num=0, label=None, dataset='', e
...
-    if input.shape[2] == 2:
+    if 1 <= input.shape[2] <= 2:
         input = input[:, :, :1]  # take first band (will become grayscale image)
-    elif input.shape[2] > 3:
+        input = np.squeeze(input)
+    elif input.shape[2] >= 3:
         input = input[:, :, :3]  # take three first bands assuming they are RGB in correct order
     mode = 'L' if input.shape[2] == 1 else 'RGB' # https://pillow.readthedocs.io/en/3.1.x/handbook/concepts.html#concept-modes
-    input_PIL = Image.fromarray(input.astype(np.uint8), mode=mode)  # FIXME: test this with grayscale input.
+    input_PIL = Image.fromarray(input.astype(np.uint8), mode=mode)  # TODO: test this with grayscale input.
```
Comment (CharlesAuthier) [inline utils/visualization.py:128]: Are we gonna have grayscale images ?

File: .travis.yml [hunk 11]
```diff
@@ -12,16 +12,17 @@ install:
...
-  - python images_to_samples.py ./conf/config_ci_segmentation_local.yaml
-  - python train_segmentation.py ./conf/config_ci_segmentation_local.yaml
-  - python inference.py ./conf/config_ci_segmentation_local.yaml
-  - python train_classification.py ./conf/config_ci_classification_local.yaml
-  - python inference.py ./conf/config_ci_classification_local.yaml
+  - python images_to_samples.py ./conf/travis_CI/config_ci_segmentation_local.yaml
+  - python train_segmentation.py ./conf/travis_CI/config_ci_segmentation_local.yaml
+  - python inference.py ./conf/travis_CI/config_ci_segmentation_local.yaml
+  - python train_classification.py ./conf/travis_CI/config_ci_classification_local.yaml
+  - python inference.py ./conf/travis_CI/config_ci_classification_local.yaml
```
Comment (victorlazio109) [inline .travis.yml:26]: travis_CI conf folder. nice!
